connman/vpn/plugins/l2tp.c

```diff
@@ -125,12 +125,40 @@ struct {
 static DBusConnection *connection;
 
 struct l2tp_private_data {
+	struct vpn_provider *provider;
 	struct connman_task *task;
 	char *if_name;
 	vpn_provider_connect_cb_t cb;
 	void *user_data;
 };
 
+static void l2tp_connect_done(struct l2tp_private_data *data, int err)
+{
+	vpn_provider_connect_cb_t cb;
+	void *user_data;
+
+	if (!data || !data->cb)
+		return;
+
+	/* Ensure that callback is called only once */
+	cb = data->cb;
+	user_data = data->user_data;
+	data->cb = NULL;
+	data->user_data = NULL;
+	cb(data->provider, user_data, err);
+}
+
+static void free_private_data(struct l2tp_private_data *data)
+{
+	if (vpn_provider_get_plugin_data(data->provider) == data)
+		vpn_provider_set_plugin_data(data->provider, NULL);
+
+	l2tp_connect_done(data, EIO);
```

Comment:
I'm not entirely sure whether I like this call, but I guess it's better safe than sorry.

Reply:
This is for the sake of binary simply crashing to make sure that the callback and D-Bus call is replied to. And other odd cases when the VPN dies because of, e.g., invalid content passed to the binary via conf file. If it is already called, then there is nothing to do. But because of varying limits for pending D-Bus calls I think this is almost a must to have here as well. Just to be sure :).

Reply:
I guess, explained? :)